### ISSUE TITLE
api: `commit` has no effect on records (and other session issues)

### ISSUE THREAD

**Issue by dset0x** [inveniosoftware/invenio-records#22]
Let's say we open a `inveniomanage shell` session:

```python
from invenio_records.api import get_record
r = get_record(1)
r['cat'] = 'dog'
r.commit()
assert r['cat'] == 'dog'
r=get_record(1)
assert r['cat'] == 'dog'
```

We close the shell and start over:
```python
from invenio_records.api import get_record
r = get_record(1)
r['cat'] 
KeyError                                  Traceback (most recent call last)
<ipython-input-3-90ff799a4799> in <module>()
----> 1 r['cat']

/home/gs-sis/.virtualenvs/invenio/src/invenio-records/invenio_records/api.pyc in __getitem__(self, key)
     43     def __getitem__(self, key):
     44         try:
---> 45             return super(Record, self).__getitem__(key)
     46         except KeyError:
     47             if key in self.__key_aliases__:

/home/gs-sis/.virtualenvs/invenio/src/invenio/invenio/utils/datastructures.pyc in __getitem__(self, key)
    255         value = self._dict
    256         for k in keys:
--> 257             value = getitem(k, value)
    258         return value
    259 

/home/gs-sis/.virtualenvs/invenio/src/invenio/invenio/utils/datastructures.pyc in getitem(k, v)
    226         def getitem(k, v):
    227             if isinstance(v, dict):
--> 228                 return v[k]
    229             elif ']' in k:
    230                 k = k[:-1].replace('n', '-1')

```
---

When using `RecordMetadata` directly I can modify the record, _but_ when trying to retrieve the data from a _separate shell that was already running before `.commit()`_, I have to restart that shell to get the modified record.

---
**Edit**: This was moved to its own issue (#24)

On a potentially related note, I somehow manage to get this exception when loading the same record twice in celery. I _never_ call `.commit()` in the code of the following script, and yet for some reason the record gets updated while being queried for:
```
Traceback (most recent call last):
  File "/home/gs-sis/.virtualenvs/invenio/src/invenio-checker/invenio_checker/conftest2.py", line 352, in record
    return get_record(request)(record_id)
  File "/home/gs-sis/.virtualenvs/invenio/src/invenio-checker/invenio_checker/conftest2.py", line 286, in _get_record
    invenio_records['original'][recid] = get_record_orig_or_mem(recid)
  File "/home/gs-sis/.virtualenvs/invenio/src/invenio-checker/invenio_checker/redis_helpers.py", line 706, in get_record_orig_or_mem
    record = get_record_orig(recid)  # This is `invenio_records.api.get_record`
  File "/home/gs-sis/.virtualenvs/invenio/src/invenio-records/invenio_records/api.py", line 125, in get_record
    obj = RecordMetadata.query.get(recid)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 819, in get
    return self._get_impl(ident, loading.load_on_ident)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 852, in _get_impl
    return fallback_fn(self, key)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/loading.py", line 219, in load_on_ident
    return q.one()
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2473, in one
    ret = list(self)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2516, in __iter__
    return self._execute_and_instances(context)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2529, in _execute_and_instances
    close_with_result=True)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2520, in _connection_from_session
    **kw)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 882, in connection
    execution_options=execution_options)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 887, in _connection_for_bind
    engine, execution_options)
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 313, in _connection_for_bind
    self._assert_active()
  File "/home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 214, in _assert_active
    % self._rollback_exception
InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (_mysql_exceptions.OperationalError) (1205, 'Lock wait timeout exceeded; try restarting transaction') [SQL: u'UPDATE record_json SET version_id=%s, json=%s WHERE record_json.id = %s AND record_json.version_id = %s'] [parameters: (12L, '{<record data>}', 2, 11L)]
```
where `<record data>` is the record dictionary.

**Comment by dset0x** [comment]
```python
from invenio_records.api import get_record
r = get_record(1)
r['cat'] = 'dog'
r.commit()
```
Database not updated.

---

```python
from invenio_records.api import get_record
r = get_record(1)
r['cat'] = 'dog'
r.commit()
from invenio.ext.sqlalchemy import db
db.session.commit()
```
Database updated!

---

This is explained by the following:
```python
from invenio_records.api import get_record
from invenio.ext.sqlalchemy import db
r = get_record(1)

db.session.dirty == IdentitySet([])
r['cat'] = 'dog'
db.session.dirty == IdentitySet([<invenio_records.models.RecordMetadata object at 0x7fe09f116310>])
```
In other words assigning to the key leaves a lingering transaction, that `commit` inside the API's `commit` does not cover.

This trasaction happens here:
```
  /tmp/script.py(5)<module>()
      1 from invenio_records.api import get_record
      2 
      3 r=get_record(3)
      4 import ipdb; ipdb.set_trace()
----> 5 r['a']='abc'

  /home/gs-sis/.virtualenvs/invenio/src/invenio-records/invenio_records/api.py(63)__setitem__()
     61                 self.__key_aliases__[key], value
     62             )
---> 63         return super(Record, self).__setitem__(key, value)
     64 
     65     def __init__(self, data, model=None):

  /home/gs-sis/.virtualenvs/invenio/src/invenio/invenio/utils/datastructures.py(267)__setitem__()
    265 
    266         if '.' not in key and ']' not in key and not extend:
--> 267             self._dict[key] = value
    268         else:
    269             keys = SmartDict.split_key_pattern.split(key)

  /home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/ext/mutable.py(645)__setitem__()
    643         """Detect dictionary set events and emit change events."""
    644         dict.__setitem__(self, key, value)
--> 645         self.changed()
    646 
    647     def setdefault(self, key, value):

  /home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/ext/mutable.py(517)changed()
    515 
    516         for parent, key in self._parents.items():
--> 517             flag_modified(parent, key)
    518 
    519     @classmethod

  /home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/attributes.py(1593)flag_modified()
   1589 
   1590     """
   1591     state, dict_ = instance_state(instance), instance_dict(instance)
   1592     impl = state.manager[key].impl
-> 1593     state._modified_event(dict_, impl, NO_VALUE, force=True)

> /home/gs-sis/.virtualenvs/invenio/local/lib/python2.7/site-packages/sqlalchemy/orm/state.py(546)_modified_event()
    544             instance_dict = self._instance_dict()
    545             if instance_dict:
--> 546                 instance_dict._modified.add(self)

```

It is important to know that on line `--> 267             self._dict[key] = value` (stack level 3 above), `self._dict` is a `<class 'sqlalchemy.ext.mutable.MutableDict'>`. This is what causes the `session` state to become dirty. However, doing one commit will not be enough, which explains snippet # 2 above.

**Comment by jirikuncar** [comment]
Using `__setitem__` on `Record` instance is not supported. I will submit patch that raises an exception until reliable solution using `jsonpatch` is found.

**Comment by dset0x** [comment]
@jirikuncar 

> Using __setitem__ on Record instance is not supported. I will submit patch that raises an exception until reliable solution using jsonpatch is found.

It seems to work with the double commit mentioned here: https://github.com/inveniosoftware/invenio-records/issues/22#issuecomment-135342911, ie:
```diff
diff --git a/invenio_records/api.py b/invenio_records/api.py
index 055526a..8e251e5 100644
--- a/invenio_records/api.py
+++ b/invenio_records/api.py
@@ -112,6 +112,7 @@ class Record(SmartDict):
 
             db.session.add(self.model)
             db.session.commit()
+            db.session.commit()
 
             toposort_send(after_record_update, self)
```

What is the canonical way of modifying records now?

**Comment by eamonnmag** [comment]
FYI @egabancho, I believe that simply replacing the ```add``` with a ```merge``` is enough to stop this error from happening.

**Comment by jirikuncar** [comment]
@eamonnmag thanks! It is not so simple though. There were issues with nesting transactions and using safepoints. @egabancho should have already a patch.

**Comment by eamonnmag** [comment]
I have a feeling after many more tests that you're right. Initially it was working for me with just the merge, but further down the line bad things started to happen again.

**Comment by lnielsen** [comment]
I think this has been fixed in latest alpha release - can I close?

**Comment by jirikuncar** [comment]
Fixed in 1.0.0aX
